Readme.md
# onOffice plugin for WordPress
![Unit tests](https://github.com/onOfficeGmbH/oo-wp-plugin/workflows/Unit%20tests/badge.svg?branch=master)

Integrate real estates, contact forms and contact persons from the onOffice Software into your WordPress website.

## Set up for development

In [./documentation/Building.md](./documentation/Building.md) you can find instructions for how to set up your development environment. There are also instructions for making a .zip file that you can upload to WordPress.

## Getting Started

1. Move the plugin directory into a new subdirectory inside the WordPress plugins directory (`wp-content/plugins/`)
2. Create a new plugin folder called `onoffice-personalized` or create a new folder inside your theme called `onoffice-theme`.
3. Copy the folder `templates.dist` to `onoffice-personalized/templates` or `onoffice-theme/templates`. This is where the newly created individual templates will go.
4. Login into your WordPress page as an administrator and go to the plugins list by navigating to `Plugins` » `Installed Plugins`. You should be able to see and activate the onOffice for WP-Websites plugin. If no API token or secret have been saved so far, a notification will show up at the top. Clicking the link will bring you to the appropriate configuration page.
5. Start editing inside the new `onoffice-personalized` or `onoffice-theme` folder.

**IMPORTANT**: Although it is safe to disable the plugin, DELETING IT WILL WIPE ALL PLUGIN-RELATED DATA FROM THE DATABASE. WE DO NOT PROVIDE ANY WARRANTY FOR DATA LOSS!

### Getting API Access

Request your own [onOffice trial version](https://onoffice.com/)

Contact us by phone (+49 241 446860) or email (support@onoffice.com) for questions concerning onOffice enterprise edition.
Please fill out this [form](https://wpplugindoc.onoffice.de/support-request/?lang=en) in regards to questions about our API or plugin development workflow.

Proceed to the next step once you have got an API token and secret.

### Configuration Basics

In comparison to other real estate WordPress-plugins, this one does not use any file transfer via FTP but from the onOffice API.
This also means, you need to enter your API credentials before configuring anything else.

#### A First Example: Creating a New Estate List
In order to create a new estate list, go to "onOffice" » "Estates" and press "Add New". **Give the new list a name**, pick the desired settings and click "Save Changes" at the bottom of the page. Going back to the estate list overview will show you the shortcode (i.e. `[oo_estate view="my new view"]`. Paste this into a new page and open the preview. The new list of estates should be embedded.

An extensive documentation can be found at [wpplugindoc.onoffice.de](https://wpplugindoc.onoffice.de/?lang=en).

## License

This project is licensed under both GNU AGPLv3 and GNU GPLv3:
 - the plugin itself is licensed under GNU AGPLv3. See LICENSE-agpl-3.0.txt.
 - config files and default templates are licensed under GNU GPLv3. See LICENSE-gpl.txt.


## Contact

onOffice GmbH\
Charlottenburger Allee 5\
52068 Aachen\
Germany

[support@onoffice.com](mailto://support@onoffice.com)\
+49 241 446860


changelog.txt
== Changelog ==

= 2.22.5 =

**Fixes**

* When there are no estates that show up on the map, there is no longer a gap in the estate list where the map would be.

= 2.22.4 =

**Fixes**

* Fixes detail view being inaccessible for reference estates.

**Changes**

* Adds a checkbox to control whether the detail pages of reference estates are accessible or should be restricted with 404 Not Found.

= 2.22.3 =

**Changes**

* The list views in forms, addresses and estates got enriched with additional information. 

= 2.22.2 =

**Changes**

* Deactivates the duplicate check for addresses created with forms.

= 2.22.1 =

**Changes**

* For the form types contact form, prospect form, owner form an additional setting that defines the contact type is introduced. Before that it was not possible to select the contact type the address is created with.

= 2.22.0 =

**Changes**

* After installing the WP-Plugin there is no longer a default form generated. This is because previously it was impossible to enter an e-mail-address in the form by default.

* Compatibility with WordPress 5.9 has been tested.

= 2.21.8 =

**Changes**

* In enterprise forms where the HTML attribute step is used the value is now set to 1 instead of 0,1 for example for all values that are floats. Users can still input decimal numbers, but is more natural for most fields.

= 2.21.7 =

**Changes**

* Users can decide in the plugin whether the references are output or not, without needing to set up a filter in onOffice.

= 2.21.6 =

**Fixes**

* Reverts the changes from release v2.21.5 due to bugs.

= 2.21.5 =

**Changes**

* The Id in the URL is now also shown like  [url]/1234-sonnige-waermende-10qm-wohnung-m/ and makes a redirect to [url]/1234-sonnige-waermende-10qm-wohnung-mit/

= 2.21.4 =

**Changes**

* All settings are bundled into a single page called "Settings" (effectively, remove the "Modules" page) with a changed order.

= 2.21.3 =

**Fixes**

* The menu entry now behaves the same way as the other default menu entries.

= 2.21.2 =

**Fixes**

* MonsterInsights plugin can now be used together with our WP plugin.

= 2.21.1 =

**Changes**

* When creating new contact forms, the option "Create address" is now checked.

= 2.21.0 =

**Changes**

* Display all of a contact person's data on the estate detail page

= 2.20.12 =

**Fixes**

* The property status is now displayed in the detail view below the property title and no longer in the fields as marketing status.

* Separator line, black hint, red hint, file and user are no longer selectable in the configuration for any list or form in the backend. (also not for addresses and estates).

= 2.20.11 =

**Changes**

* When you submit the form, the mail that is sent does now also contains  information about the  newsletter activation fields in case they were not selected in the form.

= 2.20.10 =

**Changes**

* The explanation of the option "Show title in URL" can now be translated

* In local host the error message "Please configure your API credentials first" is no longer displayed and similar estates are now shown as expected.

= 2.20.9 =

**Fixes**

* Wherever fields are used (real estate, addresses, forms) the ID of the field, which appears after the field tile is expanded, can now be selected and copied.

= 2.20.8 =

**Fixes**

* The field "vermarktungsstatus" is no longer added to the detail view when the plugin is installed.

* The saving message now also appears in case a numeric field is added to the configuration.

= 2.20.7 =

**Fixes**

* The fallback e-mail description layout has been fixed.

* The default value for radius is now displayed in the frontend porperty list.

= 2.20.6 =

**Fixes**

* Errors occured when creating or editing any form and also in the frontend on pages where the forms' shortcodes are used. The cause of the errors has been found and the errors were fixed.

* Fixed minor bugs.

= 2.20.5 =

**Changes**

* The option for indexing PDF brochures has now an understandable label which tells the user what he/ she can expact when selecting it.

* In the default_detail.php template, free texts are now displayed with line breaks.

= 2.20.4 =

**Changes**

* The plugin no longer overwrites the theme folder when updating the theme.

= 2.20.3 =

**Changes**

* The plugin is now represented by the onOffice logo.

= 2.20.2 =

**Changes**

* Contact, interest, and owner forms now all have a checkbox that configures whether an address is created. The dublicate checkbox gets a new label Check for Duplicates (override existing address if the email is the same)

= 2.20.1 =

**Changes**

* The URL now also contains id and title.

= 2.20.0 =

**Changes**

* Allow duplicating of all lists.

= 2.19.7 =

**Changes**

* When indexing for a PDF-Exposé is turned of, that's now applied to all search engines.

= 2.19.6 =

**Changes**

* Adjusted the hint text, label and position of the email-address field.

= 2.19.5 =

**Changes**

* Configuration of fallback email address is now required in the settings for contact forms.

= 2.19.4 =

**Changes**

* Removal of the field options "Filterable", "Hidden" and "Reduce values accordings to selected filter" in unit list settings.

= 2.19.3 =

* Removal of contact person's data (Ansprechpartner der Einheit) in unit list.

= 2.19.2 =

**Changes**

* Add missing code from release 2.19.1

= 2.19.1 =

**Changes**

* The estate status is now also displayed in the detail view.

= 2.19.0 =

**Changes**

* WP-Plugin - communication with the onOffice API only takes place in the relevant cases

= 2.18.2 =

**Fixes**

* Fix number fields' default values allowing text

= 2.18.1 =

**Fixes**

* Fix for the radius search of similar properties

= 2.18.0 =

**New**

* The templates and translations can now be located in a new folder called `onoffice-theme/languages` and `onoffice-theme/templates` inside the WP theme, respectively.

= 2.17.1 =

**Fixes**


plugin.php
<?php

/**
 *
 *    Copyright (C) 2017-2019 onOffice GmbH
 *
 *    This program is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU Affero General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    This program is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU Affero General Public License for more details.
 *
 *    You should have received a copy of the GNU Affero General Public License
 *    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

/*
Plugin Name: onOffice for WP-Websites
Plugin URI: https://wpplugindoc.onoffice.de
Author: onOffice GmbH
Author URI: https://en.onoffice.com/
Description: Your connection to onOffice: This plugin enables you to have quick access to estates and forms – no additional sync with the software is needed. Consult support@onoffice.de for source code.
Version: 3.0
License: AGPL 3+
License URI: https://www.gnu.org/licenses/agpl-3.0
Text Domain: onoffice-for-wp-websites
Domain Path: /languages
*/

defined( 'ABSPATH' ) or die();

require __DIR__ . '/vendor/autoload.php';

define('ONOFFICE_PLUGIN_DIR', __DIR__);

use DI\ContainerBuilder;
use onOffice\WPlugin\Cache\CachedOutput;
use onOffice\WPlugin\Cache\CacheHandler;
use onOffice\WPlugin\Controller\AdminViewController;
use onOffice\WPlugin\Controller\ContentFilter\ContentFilterShortCodeRegistrator;
use onOffice\WPlugin\Controller\DetailViewPostSaveController;
use onOffice\WPlugin\Controller\EstateViewDocumentTitleBuilder;
use onOffice\WPlugin\Controller\RewriteRuleBuilder;
use onOffice\WPlugin\DataView\DataDetailViewCheckAccessControl;
use onOffice\WPlugin\DataView\DataDetailViewHandler;
use onOffice\WPlugin\Factory\EstateListFactory;
use onOffice\WPlugin\Field\EstateKindTypeReader;
use onOffice\WPlugin\Form\CaptchaDataChecker;
use onOffice\WPlugin\Form\Preview\FormPreviewApplicantSearch;
use onOffice\WPlugin\Form\Preview\FormPreviewEstate;
use onOffice\WPlugin\FormPostHandler;
use onOffice\WPlugin\Installer\DatabaseChangesInterface;
use onOffice\WPlugin\Installer\Installer;
use onOffice\WPlugin\PDF\PdfDocumentModel;
use onOffice\WPlugin\PDF\PdfDocumentModelValidationException;
use onOffice\WPlugin\PDF\PdfDownload;
use onOffice\WPlugin\PDF\PdfDownloadException;
use onOffice\WPlugin\Record\EstateIdRequestGuard;
use onOffice\WPlugin\ScriptLoader\ScriptLoaderRegistrator;
use onOffice\WPlugin\Controller\EstateDetailUrl;
use onOffice\WPlugin\Utility\__String;
use onOffice\WPlugin\WP\WPQueryWrapper;

define('ONOFFICE_DI_CONFIG_PATH', implode(DIRECTORY_SEPARATOR, [ONOFFICE_PLUGIN_DIR, 'config', 'di-config.php']));

$pDIBuilder = new ContainerBuilder();
$pDIBuilder->addDefinitions(ONOFFICE_DI_CONFIG_PATH);
$pDI = $pDIBuilder->build();

$pAdminViewController = new AdminViewController();
$pDetailViewPostSaveController = $pDI->get(DetailViewPostSaveController::class);
$pDI->get(ScriptLoaderRegistrator::class)->generate();

add_action('plugins_loaded', function() use ($pDI) {
	$pDI->get(DatabaseChangesInterface::class)->install();
});

add_action('init', function() use ($pDI) {
	$pRewriteRuleBuilder = $pDI->get(RewriteRuleBuilder::class);
	$pRewriteRuleBuilder->addCustomRewriteTags();
	$pRewriteRuleBuilder->addStaticRewriteRules();
	$pRewriteRuleBuilder->addDynamicRewriteRules();
});

// This hook [wp] is one effective place to perform any high-level filtering or validation,
// following queries, but before WordPress does any routing, processing, or handling.
// https://codex.wordpress.org/Plugin_API/Action_Reference/wp
add_action('wp', [FormPostHandler::class, 'initialCheck']);

add_action('admin_menu', [$pAdminViewController, 'register_menu']);
add_action('admin_enqueue_scripts', [$pAdminViewController, 'enqueue_ajax']);
add_action('admin_enqueue_scripts', [$pAdminViewController, 'enqueue_css']);
add_action('admin_enqueue_scripts', [$pAdminViewController, 'enqueueExtraJs']);
add_action('admin_notices', [$pAdminViewController, 'generalAdminNoticeSEO']);
add_action('wp_enqueue_scripts', [CaptchaDataChecker::class, 'registerScripts']);
add_action('save_post', [$pDetailViewPostSaveController, 'onSavePost']);
add_action('wp_trash_post', [$pDetailViewPostSaveController, 'onMoveTrash']);
add_action('oo_cache_cleanup', function() use ($pDI) {
	$pDI->get(CacheHandler::class)->clean();
});

add_action('init', [$pAdminViewController, 'onInit']);
add_action('init', function() use ($pAdminViewController) {
	$pAdminViewController->disableHideMetaboxes();
}, 11);
add_action('admin_init', [$pAdminViewController, 'add_ajax_actions']);
add_action('admin_init', [CaptchaDataChecker::class, 'addHook']);
add_action('admin_init', [$pDetailViewPostSaveController, 'getAllPost']);
add_action('plugins_loaded', function() {
	load_plugin_textdomain('onoffice-for-wp-websites', false, basename(ONOFFICE_PLUGIN_DIR) . '/languages');
	// Check 'onoffice-personalized' Folder exists
	$onofficePersonalizedFolderLanguages = plugin_dir_path(__DIR__) . 'onoffice-personalized/languages';
	$onofficePersonalizedFolder = plugin_dir_path(__DIR__) . 'onoffice-personalized';
	$onofficeThemeFolderLanguages = get_template_directory() . '/onoffice-theme/languages';

	if (is_dir($onofficeThemeFolderLanguages)) {
		load_textdomain('onoffice', $onofficeThemeFolderLanguages . '/onoffice-'.get_locale().'.mo');
	} elseif (is_dir($onofficePersonalizedFolderLanguages)) {
		load_plugin_textdomain('onoffice', false, basename($onofficePersonalizedFolder) . '/languages');
	} else {
		load_plugin_textdomain('onoffice', false, basename(ONOFFICE_PLUGIN_DIR) . '/languages');
	}
});

// "Settings" link in plugins list
add_filter('plugin_action_links_'.plugin_basename(__FILE__), [$pAdminViewController, 'pluginSettingsLink']);

$pDI->get(ContentFilterShortCodeRegistrator::class)->register();

add_filter('document_title_parts', function($title) use ($pDI) {
	return $pDI->get(EstateViewDocumentTitleBuilder::class)->buildDocumentTitle($title);
}, 10, 2);
if (get_option('onoffice-settings-title-and-description') === '1')
{
	add_filter('get_post_metadata', function($value, $object_id, $meta_key) use ($pDI) {
			$pDataDetailViewHandler = $pDI->get( DataDetailViewHandler::class );
			$pDetailView = $pDataDetailViewHandler->getDetailView();
			$detail_page_id = $pDetailView->getPageId();
			if ( $object_id == $detail_page_id ) {
				$list_meta_keys = [
					"onoffice_titel"            => 'objekttitel',
					"onoffice_title"            => 'objekttitel',
					"onoffice_beschreibung"     => 'objektbeschreibung',
					"onoffice_description"      => 'objektbeschreibung',
					"onoffice_ort"              => 'ort',
					"onoffice_city"             => 'ort',
					"onoffice_plz"              => 'plz',
					"onoffice_postal_code"      => 'plz',
					"onoffice_objektart"        => 'objektart',
					"onoffice_property_class"   => 'objektart',
					"onoffice_vermarktungsart"  => 'vermarktungsart',
					"onoffice_marketing_method" => 'vermarktungsart',
					"onoffice_datensatznr"      => 'Id',
					"onoffice_id"               => 'Id'
				];
				if ( isset( $list_meta_keys[ $meta_key ] ) ) {
					return customFieldCallback( $pDI, $list_meta_keys[ $meta_key ] );
				}
			} else {
				return null;
			}
	}, 1, 3);
}

// Return title custom by custom field onOffice
function customFieldCallback( $pDI, $format) {
	return $pDI->get( EstateViewDocumentTitleBuilder::class )->buildDocumentTitleField($format);
}


add_filter('wpml_ls_language_url', function($url) use ($pDI){
	$pWPQueryWrapper = $pDI->get(WPQueryWrapper::class);
	$estateId = (int) $pWPQueryWrapper->getWPQuery()->get('estate_id', 0);
	return $pDI->get(EstateDetailUrl::class)->createEstateDetailLink($url, $estateId);
}, 10, 2);

register_activation_hook(__FILE__, [Installer::class, 'install']);
register_deactivation_hook(__FILE__, [Installer::class, 'deactivate']);
register_uninstall_hook(__FILE__, [Installer::class, 'deinstall']);

if (!wp_next_scheduled('oo_cache_cleanup')) {
	wp_schedule_event(time(), 'hourly', 'oo_cache_cleanup');
}

// Gets triggered before we know if it has to be updated at all, so that no value has to be changed
add_action('pre_update_option', function($value, $option) use ($pDI) {
	if (__String::getNew($option)->startsWith('onoffice')) {
		$pDI->get(CacheHandler::class)->clear();
	}
	return $value;
}, 10, 2);

add_filter('query_vars', function(array $query_vars): array {
    $query_vars []= 'onoffice_estate_type_json';
    $query_vars []= 'onoffice_applicant_search_preview';
    $query_vars []= 'onoffice_estate_preview';
    $query_vars []= 'document_pdf';
    $query_vars []= 'preview_name';
    $query_vars []= 'nonce';
    return $query_vars;
});

add_action('parse_request', function(WP $pWP) use ($pDI) {
	if (isset($pWP->query_vars['document_pdf'])) {
		try {
			$pPdfDocumentModel = new PdfDocumentModel($pWP->query_vars['estate_id'] ?? 0, $pWP->query_vars['view'] ?? '');
			/* @var $pPdfDownload PdfDownload */
			$pPdfDownload = $pDI->get(PdfDownload::class);
			$pPdfDownload->download($pPdfDocumentModel);
		} catch (PdfDocumentModelValidationException $pEx) {
			$pWP->handle_404();
			include(get_query_template('404'));
		} catch (PdfDownloadException $pEx) {
			$pWP->handle_404();
			include(get_query_template('404'));
		}
		die();
	}
});

add_action('parse_request', function(WP $pWP) use ($pDI) {
	$estateId = $pWP->query_vars['estate_id'] ?? '';
	/** @var EstateIdRequestGuard $pEstateIdGuard */
	$pEstateIdGuard = $pDI->get(EstateIdRequestGuard::class);
	/** @var DataDetailViewHandler $pDataDetailViewHandler */
	$pDataDetailViewHandler = $pDI->get( DataDetailViewHandler::class);

	if ($estateId !== '') {
		$estateId = (int)$estateId;
		/** @var DataDetailViewCheckAccessControl $pDataDetailViewCheckAccessControl */
		$pDataDetailViewCheckAccessControl = $pDI->get(DataDetailViewCheckAccessControl::class);
		$accessControlChecker = $pDataDetailViewCheckAccessControl->checkAccessControl($estateId);

		if ($estateId === 0 || !$accessControlChecker|| !$pEstateIdGuard->isValid($estateId)) {
			$pWP->handle_404();
			include(get_query_template('404'));
			die();
		}
	}
});

add_action('parse_request', static function(WP $pWP) use ($pDI) {
	if (isset($pWP->query_vars['onoffice_estate_type_json'])) {
		$content = wp_json_encode($pDI->get(EstateKindTypeReader::class)->read());
		/** @var CachedOutput $pCachedOutput */
		$pCachedOutput = $pDI->get(CachedOutput::class);
		header('Content-Type: application/json; charset='.get_option('blog_charset'));
		$pCachedOutput->outputCached($content, 60 * 60 * 24 * 2);
		exit;
	}
});

add_action('parse_request', function(WP $pWP) use ($pDI) {
	if (isset($pWP->query_vars['onoffice_estate_preview'], $pWP->query_vars['preview_name']) &&
		wp_verify_nonce($pWP->query_vars['nonce'], 'onoffice-estate-preview') === 1) {
		wp_send_json($pDI->get(FormPreviewEstate::class)
			->preview((string)$pWP->query_vars['preview_name']));
	}
});

add_action('parse_request', function(WP $pWP) use ($pDI) {
	if (isset($pWP->query_vars['onoffice_applicant_search_preview'], $pWP->query_vars['preview_name']) &&
		wp_verify_nonce($pWP->query_vars['nonce'], 'onoffice-applicant-search-preview') === 1) {
		wp_send_json($pDI->get(FormPreviewApplicantSearch::class)
			->preview((string)$pWP->query_vars['preview_name']));
	}
});

add_filter('set-screen-option', function ($status, $option, $value) {
	$pagination_screen_option = array(
		"onoffice_address_listview_per_page",
		"onoffice_estate_listview_per_page",
		"onoffice_forms_forms_per_page",
		"onoffice_estate_units_listview_per_page"
	);

	if (in_array($option, $pagination_screen_option)) {
		return $value;
	}
	return $status;
}, 10, 3);

function update_duplicate_check_warning_option()
{
	update_option('onoffice-duplicate-check-warning', 0);
	echo true;
	wp_die();
}

add_action('wp_ajax_update_duplicate_check_warning_option', 'update_duplicate_check_warning_option');

return $pDI;
